File: TeamProject/src/main/java/org/coms4156/tars/client/TarsApiClient.java
package org.coms4156.tars.client;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;
import org.coms4156.tars.model.CitySummary;
import org.coms4156.tars.model.CrimeSummary;
import org.coms4156.tars.model.TarsUser;
import org.coms4156.tars.model.User;
import org.coms4156.tars.model.WeatherAlert;
import org.coms4156.tars.model.WeatherRecommendation;
import org.springframework.stereotype.Component;

/**
 * Client for interacting with the TARS API.
 * Provides methods to call all available endpoints.
 */
@Component
public class TarsApiClient {
  private final String baseUrl;
  private final HttpClient httpClient;
  private final ObjectMapper objectMapper;

  /**
   * Creates a new TarsApiClient with the default base URL (http://localhost:8080).
   */
  public TarsApiClient() {
    this("http://localhost:8080");
  }

  /**
   * Creates a new TarsApiClient with a custom base URL.
   *
   * @param baseUrl the base URL of the TARS API (e.g., "http://localhost:8080")
   */
  public TarsApiClient(String baseUrl) {
    this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
    this.httpClient = HttpClient.newHttpClient();
    this.objectMapper = new ObjectMapper();
  }

  /**
   * Gets the welcome message from the index endpoint.
   *
   * @return the welcome message string
   * @throws IOException if an I/O error occurs
   * @throws InterruptedException if the request is interrupted
   */
  public String getIndex() throws IOException, InterruptedException {
    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(baseUrl + "/"))
        .GET()
        .build();

    HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    return response.body();
  }

  /**
   * Creates a new client.
   *
   * @param name the client name
   * @param email the client email
   * @return a map containing the response data
   * @throws IOException if an I/O error occurs
   * @throws InterruptedException if the request is interrupted
   */
  public Map<String, Object> createClient(String name, String email)
      throws IOException, InterruptedException {
    Map<String, String> body = Map.of("name", name, "email", email);
    String jsonBody = objectMapper.writeValueAsString(body);

    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(baseUrl + "/client/create"))
        .header("Content-Type", "application/json")
        .POST(HttpRequest.BodyPublishers.ofString(jsonBody))
        .build();

    HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    return objectMapper.readValue(response.body(), new TypeReference<Map<String, Object>>() {});
  }

  /**
   * Creates a new user for a client.
   *
   * @param clientId the client ID
   * @param username the username
   * @param email the user email
   * @param role the user role
   * @return the created TarsUser
   * @throws IOException if an I/O error occurs
   * @throws InterruptedException if the request is interrupted
   */
  public TarsUser createClientUser(Long clientId, String username, String email, String role)
      throws IOException, InterruptedException {
    TarsUser user = new TarsUser(clientId, username, email, role);
    String jsonBody = objectMapper.writeValueAsString(user);

    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(baseUrl + "/client/createUser"))
        .header("Content-Type", "application/json")
        .POST(HttpRequest.BodyPublishers.ofString(jsonBody))
        .build();

    HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    return objectMapper.readValue(response.body(), TarsUser.class);
  }

  /**
   * Adds user preferences for a user.
   *
   * @param userId the user ID
   * @param user the User object with preferences
   * @return the User object with preferences
   * @throws IOException if an I/O error occurs
   * @throws InterruptedException if the request is interrupted
   */
  public User addUser(int userId, User user) throws IOException, InterruptedException {
    String jsonBody = objectMapper.writeValueAsString(user);

    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(baseUrl + "/user/" + userId + "/add"))
        .header("Content-Type", "application/json")
        .PUT(HttpRequest.BodyPublishers.ofString(jsonBody))
        .build();

    HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    return objectMapper.readValue(response.body(), User.class);
  }

  /**
   * Updates user preferences for an existing user.
   *
   * @param userId the user ID
   * @param user the User object with updated preferences
   * @return the updated User object with preferences
   * @throws IOException if an I/O error occurs
   * @throws InterruptedException if the request is interrupted
   */
  public User updateUser(int userId, User user) throws IOException, InterruptedException {
    String jsonBody = objectMapper.writeValueAsString(user);

    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(baseUrl + "/user/" + userId + "/update"))
        .header("Content-Type", "application/json")
        .PUT(HttpRequest.BodyPublishers.ofString(jsonBody))
        .build();

    HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    
    // Check if response is successful
    if (response.statusCode() >= 200 && response.statusCode() < 300) {
      return objectMapper.readValue(response.body(), User.class);
    } else {
      throw new IOException("Failed to update user: " + response.body());
    }
  }

  /**
   * Gets user preferences by user ID.
   *
   * @param userId the user ID
   * @return the User object with preferences
   * @throws IOException if an I/O error occurs
   * @throws InterruptedException if the request is interrupted
   */
  public User getUser(int userId) throws IOException, InterruptedException {
    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(baseUrl + "/user/" + userId))
        .GET()
        .build();

    HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    return objectMapper.readValue(response.body(), User.class);
  }

  /**
   * Gets the list of all users.
   *
   * @return a list of all User objects
   * @throws IOException if an I/O error occurs
   * @throws InterruptedException if the request is interrupted
   */
  public List<User> getUserList() throws IOException, InterruptedException {
    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(baseUrl + "/userList"))
        .GET()
        .build();

    HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    return objectMapper.readValue(response.body(),
        new TypeReference<List<User>>() {});
  }

  /**
   * Gets weather recommendations for a city.
   *
   * @param city the city name
   * @param days the number of forecast days (1-14)
   * @return the WeatherRecommendation
   * @throws IOException if an I/O error occurs
   * @throws InterruptedException if the request is interrupted
   */
  public WeatherRecommendation getWeatherRecommendation(String city, int days)
      throws IOException, InterruptedException {
    String encodedCity = URLEncoder.encode(city, StandardCharsets.UTF_8);
    String uri = baseUrl + "/recommendation/weather/?city=" + encodedCity + "&days=" + days;

    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(uri))
        .GET()
        .build();

    HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    return objectMapper.readValue(response.body(), WeatherRecommendation.class);
  }

  /**
   * Gets weather alerts by city name.
   *
   * @param city the city name
   * @return the WeatherAlert
   * @throws IOException if an I/O error occurs
   * @throws InterruptedException if the request is interrupted
   */
  public WeatherAlert getWeatherAlertsByCity(String city)
      throws IOException, InterruptedException {
    String encodedCity = URLEncoder.encode(city, StandardCharsets.UTF_8);
    String uri = baseUrl + "/alert/weather?city=" + encodedCity;

    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(uri))
        .GET()
        .build();

    HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    return objectMapper.readValue(response.body(), WeatherAlert.class);
  }

  /**
   * Gets weather alerts by coordinates.
   *
   * @param lat the latitude
   * @param lon the longitude
   * @return the WeatherAlert
   * @throws IOException if an I/O error occurs
   * @throws InterruptedException if the request is interrupted
   */
  public WeatherAlert getWeatherAlertsByCoordinates(double lat, double lon)
      throws IOException, InterruptedException {
    String uri = baseUrl + "/alert/weather?lat=" + lat + "&lon=" + lon;

    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(uri))
        .GET()
        .build();

    HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    return objectMapper.readValue(response.body(), WeatherAlert.class);
  }

  /**
   * Gets weather alerts for a specific user based on their city preferences.
   *
   * @param userId the user ID
   * @return a list of WeatherAlert objects
   * @throws IOException if an I/O error occurs
   * @throws InterruptedException if the request is interrupted
   */
  public List<WeatherAlert> getUserWeatherAlerts(int userId)
      throws IOException, InterruptedException {
    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(baseUrl + "/alert/weather/user/" + userId))
        .GET()
        .build();

    HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    return objectMapper.readValue(response.body(),
        new TypeReference<List<WeatherAlert>>() {});
  }

  /**
   * Gets crime summary data.
   *
   * @param state the state name or abbreviation
   * @param offense the offense type
   * @param month the month (two-digit format, e.g., "10")
   * @param year the year (four-digit format, e.g., "2025")
   * @return the CrimeSummary
   * @throws IOException if an I/O error occurs
   * @throws InterruptedException if the request is interrupted
   */
  public CrimeSummary getCrimeSummary(String state, String offense, String month, String year)
      throws IOException, InterruptedException {
    String encodedState = URLEncoder.encode(state, StandardCharsets.UTF_8);
    String encodedOffense = URLEncoder.encode(offense, StandardCharsets.UTF_8);
    String encodedMonth = URLEncoder.encode(month, StandardCharsets.UTF_8);
    String encodedYear = URLEncoder.encode(year, StandardCharsets.UTF_8);
    String uri = baseUrl + "/crime/summary?state=" + encodedState
        + "&offense=" + encodedOffense
        + "&month=" + encodedMonth
        + "&year=" + encodedYear;

    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(uri))
        .GET()
        .build();

    HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    return objectMapper.readValue(response.body(), CrimeSummary.class);
  }

  /**
   * Gets travel advisory for a country.
   *
   * @param country the country name
   * @return the travel advisory as a string
   * @throws IOException if an I/O error occurs
   * @throws InterruptedException if the request is interrupted
   */
  public String getCountryAdvisory(String country)
      throws IOException, InterruptedException {
    String encodedCountry = URLEncoder.encode(country, StandardCharsets.UTF_8);
    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(baseUrl + "/country/" + encodedCountry))
        .GET()
        .build();

    HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    return response.body();
  }

  /**
   * Gets city summary for a city.
   *
   * @param city the city name
   * @return the CitySummary
   * @throws IOException if an I/O error occurs
   * @throws InterruptedException if the request is interrupted
   */
  public CitySummary getCitySummary(String city)
      throws IOException, InterruptedException {
    return getCitySummary(city, null, null);
  }

  /**
   * Gets city summary for a city with optional date range.
   *
   * @param city the city name
   * @param startDate the start date (optional, format: YYYY-MM-DD)
   * @param endDate the end date (optional, format: YYYY-MM-DD)
   * @return the CitySummary
   * @throws IOException if an I/O error occurs
   * @throws InterruptedException if the request is interrupted
   */
  public CitySummary getCitySummary(String city, String startDate, String endDate)
      throws IOException, InterruptedException {
    String encodedCity = URLEncoder.encode(city, StandardCharsets.UTF_8);
    StringBuilder uriBuilder = new StringBuilder(baseUrl + "/summary/" + encodedCity);

    if (startDate != null || endDate != null) {
      uriBuilder.append("?");
      if (startDate != null) {
        uriBuilder.append("startDate=").append(URLEncoder.encode(startDate, StandardCharsets.UTF_8));
      }
      if (endDate != null) {
        if (startDate != null) {
          uriBuilder.append("&");
        }
        uriBuilder.append("endDate=").append(URLEncoder.encode(endDate, StandardCharsets.UTF_8));
      }
    }

    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(uriBuilder.toString()))
        .GET()
        .build();

    HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    return objectMapper.readValue(response.body(), CitySummary.class);
  }
}

File: tars-client/src/tarsApiClient.js
const axios = require('axios');
const fs = require('fs-extra');
const path = require('path');

/**
 * TARS API Client for Node.js
 * Wraps the Java TARS API endpoints and manages persistent client ID storage
 */
class TarsApiClient {
  constructor(baseUrl = 'http://localhost:8080') {
    this.baseUrl = baseUrl.endsWith('/') ? baseUrl.slice(0, -1) : baseUrl;
    this.clientConfigPath = path.join(__dirname, '..', 'client-config.json');
    this.clientId = null;
    this.loadClientId();
  }

  /**
   * Load client ID from local storage
   */
  loadClientId() {
    try {
      if (fs.existsSync(this.clientConfigPath)) {
        const config = fs.readJsonSync(this.clientConfigPath);
        this.clientId = config.clientId;
        console.log(`Loaded client ID: ${this.clientId}`);
      }
    } catch (error) {
      console.warn('Could not load client ID:', error.message);
    }
  }

  /**
   * Save client ID to local storage
   */
  saveClientId(clientId) {
    try {
      this.clientId = clientId;
      fs.writeJsonSync(this.clientConfigPath, { clientId }, { spaces: 2 });
      console.log(`Saved client ID: ${clientId}`);
    } catch (error) {
      console.error('Could not save client ID:', error.message);
    }
  }

  /**
   * Get existing client ID or create a new one
   */
  async getOrCreateClientId() {
    if (this.clientId) {
      return this.clientId;
    }

    // Try to create a new client with a unique name
    const timestamp = Date.now();
    const name = `Client-${timestamp}`;
    const email = `client-${timestamp}@tars.local`;

    try {
      const result = await this.createClient(name, email);
      if (result && result.clientId) {
        this.saveClientId(result.clientId);
        return result.clientId;
      }
    } catch (error) {
      console.error('Failed to create client:', error.message);
      throw error;
    }

    throw new Error('Failed to get or create client ID');
  }

  /**
   * Get the welcome message from the index endpoint
   */
  async getIndex() {
    try {
      const response = await axios.get(`${this.baseUrl}/`);
      return response.data;
    } catch (error) {
      throw new Error(`Failed to get index: ${error.message}`);
    }
  }

  /**
   * Create a new client
   */
  async createClient(name, email) {
    try {
      const response = await axios.post(`${this.baseUrl}/client/create`, {
        name,
        email
      });
      return response.data;
    } catch (error) {
      const errorMsg = error.response?.data 
        ? (typeof error.response.data === 'object' 
            ? JSON.stringify(error.response.data) 
            : error.response.data)
        : error.message;
      throw new Error(`Failed to create client: ${errorMsg}`);
    }
  }

  /**
   * Create a new user for a client
   */
  async createClientUser(clientId, username, email, role) {
    try {
      const response = await axios.post(`${this.baseUrl}/client/createUser`, {
        clientId,
        username,
        email,
        role
      });
      return response.data;
    } catch (error) {
      const errorMsg = error.response?.data 
        ? (typeof error.response.data === 'object' 
            ? JSON.stringify(error.response.data) 
            : error.response.data)
        : error.message;
      throw new Error(`Failed to create user: ${errorMsg}`);
    }
  }

  /**
   * Set user preferences (adds or updates)
   */
  async setUserPreference(userId, user) {
    try {
      console.log(`[tarsApiClient] Setting preference for userId: ${userId}`, user);
      const response = await axios.put(`${this.baseUrl}/setPreference/${userId}`, user);
      console.log(`[tarsApiClient] Success response:`, response.data);
      return response.data;
    } catch (error) {
      console.error(`[tarsApiClient] Error setting preference:`, error);
      console.error(`[tarsApiClient] Error response:`, error.response);
      // Extract error message from response
      let errorMsg = error.message;
      if (error.response?.data) {
        if (typeof error.response.data === 'string') {
          errorMsg = error.response.data;
        } else if (typeof error.response.data === 'object') {
          errorMsg = error.response.data.error || JSON.stringify(error.response.data);
        }
      }
      // Preserve the original error structure so server.js can access it
      const enhancedError = error; // Keep the original error
      enhancedError.message = errorMsg; // Update message
      throw enhancedError;
    }
  }

  /**
   * Remove user
   */
  async removeUser(userId) {
    try {
      const response = await axios.put(`${this.baseUrl}/user/${userId}/remove`);
      return response.data;
    } catch (error) {
      throw new Error(`Failed to remove user: ${error.response?.data || error.message}`);
    }
  }

  /**
   * Get user by ID
   */
  async getUser(userId) {
    try {
      const response = await axios.get(`${this.baseUrl}/user/${userId}`);
      return response.data;
    } catch (error) {
      throw new Error(`Failed to get user: ${error.response?.data || error.message}`);
    }
  }

  /**
   * Get all users
   */
  async getUserList() {
    try {
      const response = await axios.get(`${this.baseUrl}/userList`);
      return response.data;
    } catch (error) {
      throw new Error(`Failed to get user list: ${error.response?.data || error.message}`);
    }
  }

  /**
   * Get users for a specific client
   */
  async getClientUserList(clientId) {
    try {
      const response = await axios.get(`${this.baseUrl}/userList/client/${clientId}`);
      return response.data;
    } catch (error) {
      throw new Error(`Failed to get client user list: ${error.response?.data || error.message}`);
    }
  }

  /**
   * Get all TarsUsers from users.json
   */
  async getTarsUsers() {
    try {
      const response = await axios.get(`${this.baseUrl}/tarsUsers`);
      return response.data;
    } catch (error) {
      throw new Error(`Failed to get TARS users: ${error.response?.data || error.message}`);
    }
  }

  /**
   * Delete a TarsUser by userId
   */
  async deleteTarsUser(userId) {
    try {
      const response = await axios.delete(`${this.baseUrl}/tarsUsers/${userId}`);
      return response.data;
    } catch (error) {
      throw new Error(`Failed to delete TARS user: ${error.response?.data || error.message}`);
    }
  }

  /**
   * Get weather recommendation
   */
  async getWeatherRecommendation(city, days) {
    try {
      const response = await axios.get(`${this.baseUrl}/recommendation/weather`, {
        params: { city, days }
      });
      return response.data;
    } catch (error) {
      throw new Error(`Failed to get weather recommendation: ${error.response?.data || error.message}`);
    }
  }

  /**
   * Get weather recommendation for a user based on their preferences (getUserRec)
   * userId is now a path variable for consistency
   */
  async getUserWeatherRecommendation(city, userId, days) {
    try {
      const response = await axios.get(`${this.baseUrl}/recommendation/weather/user/${userId}`, {
        params: { city, days }
      });
      return response.data;
    } catch (error) {
      throw new Error(`Failed to get user weather recommendation: ${error.response?.data || error.message}`);
    }
  }

  /**
   * Get weather alerts by city
   */
  async getWeatherAlertsByCity(city) {
    try {
      const response = await axios.get(`${this.baseUrl}/alert/weather`, {
        params: { city }
      });
      return response.data;
    } catch (error) {
      throw new Error(`Failed to get weather alerts: ${error.response?.data || error.message}`);
    }
  }

  /**
   * Get weather alerts by coordinates
   */
  async getWeatherAlertsByCoordinates(lat, lon) {
    try {
      const response = await axios.get(`${this.baseUrl}/alert/weather`, {
        params: { lat, lon }
      });
      return response.data;
    } catch (error) {
      throw new Error(`Failed to get weather alerts: ${error.response?.data || error.message}`);
    }
  }

  /**
   * Get user weather alerts
   */
  async getUserWeatherAlerts(userId) {
    try {
      const response = await axios.get(`${this.baseUrl}/alert/weather/user/${userId}`);
      return response.data;
    } catch (error) {
      throw new Error(`Failed to get user weather alerts: ${error.response?.data || error.message}`);
    }
  }

  /**
   * Get crime summary
   */
  async getCrimeSummary(state, offense, month, year) {
    try {
      const response = await axios.get(`${this.baseUrl}/crime/summary`, {
        params: { 
          state: String(state), 
          offense: String(offense), 
          month: String(month), 
          year: String(year) 
        }
      });
      return response.data;
    } catch (error) {
      const errorMsg = error.response?.data?.message || error.response?.data || error.message;
      throw new Error(`Failed to get crime summary: ${errorMsg}`);
    }
  }

  /**
   * Get country advisory
   */
  async getCountryAdvisory(country) {
    try {
      const response = await axios.get(`${this.baseUrl}/country/${encodeURIComponent(country)}`);
      return response.data;
    } catch (error) {
      throw new Error(`Failed to get country advisory: ${error.response?.data || error.message}`);
    }
  }

  /**
   * Get country summary
   */
  async getCountrySummary(country) {
    try {
      const response = await axios.get(
        `${this.baseUrl}/countrySummary/${encodeURIComponent(country)}`
      );
      return response.data;
    } catch (error) {
      const errorMsg = error.response?.data 
        ? (typeof error.response.data === 'object' 
            ? JSON.stringify(error.response.data) 
            : error.response.data)
        : error.message;
      throw new Error(`Failed to get country summary: ${errorMsg}`);
    }
  }

  /**
   * Get all clients
   */
  async getClients() {
    try {
      const response = await axios.get(`${this.baseUrl}/clients`);
      return response.data;
    } catch (error) {
      const errorMsg = error.response?.data 
        ? (typeof error.response.data === 'object' 
            ? JSON.stringify(error.response.data) 
            : error.response.data)
        : error.message;
      throw new Error(`Failed to get clients: ${errorMsg}`);
    }
  }

  /**
   * Get user by clientId
   */
  async getUserByClientId(clientId) {
    try {
      const response = await axios.get(`${this.baseUrl}/user/client/${clientId}`);
      return response.data;
    } catch (error) {
      const errorMsg = error.response?.data 
        ? (typeof error.response.data === 'object' 
            ? JSON.stringify(error.response.data) 
            : error.response.data)
        : error.message;
      throw new Error(`Failed to get user by clientId: ${errorMsg}`);
    }
  }

  /**
   * Get user preferences by userId
   */
  async getUserPreference(userId) {
    try {
      const response = await axios.get(`${this.baseUrl}/retrievePreference/${userId}`);
      return response.data;
    } catch (error) {
      const errorMsg = error.response?.data 
        ? (typeof error.response.data === 'object' 
            ? JSON.stringify(error.response.data) 
            : error.response.data)
        : error.message;
      throw new Error(`Failed to get user preference: ${errorMsg}`);
    }
  }

  /**
   * Login user by username, email, or userId
   */
  async login(username, email, userId) {
    try {
      const body = {};
      if (username) body.username = username;
      if (email) body.email = email;
      if (userId) body.userId = userId;
      
      const response = await axios.post(`${this.baseUrl}/login`, body);
      return response.data;
    } catch (error) {
      const errorMsg = error.response?.data 
        ? (typeof error.response.data === 'object' 
            ? JSON.stringify(error.response.data) 
            : error.response.data)
        : error.message;
      throw new Error(`Failed to login: ${errorMsg}`);
    }
  }
}

module.exports = TarsApiClient;

